syntax = "proto3";

import "beams/service/remote_calls/generic_message.proto"; // MESSAGE_TYPE_BEHAVIOR_TREE_MESSAGE import

/* Behavior Tree Update related fields */

// Names must match py_trees.common.Status fields precisely
enum TickStatus {
  INVALID = 0;
  RUNNING = 1;
  SUCCESS = 2;
  FAILURE = 3;
}

enum TickConfiguration {
  UNKNOWN = 0;
  INTERACTIVE = 1;
  CONTINUOUS = 2;
}

enum TreeStatus {
  IDLE = 0;
  TICKING = 1;
  WAITING_ACK = 2;
  ERROR = 3;
}

// identification information for a node
message NodeId {
  string name = 1;
  string uuid = 2;
}

message BehaviorTreeUpdateMessage {
  MessageType mess_t = 1; // MESSAGE_TYPE_BEHAVIOR_TREE_MESSAGE
  NodeId tree_id = 2;
  NodeId node_id = 3;
  TickStatus tick_status = 4;
  TickConfiguration tick_config = 5;
  int32 tick_delay_ms = 6;
  TreeStatus tree_status = 7;
}

message NodeInfo {
  NodeId id = 1;
  TickStatus status = 2;
  repeated NodeInfo children = 3;
}

message TreeDetails {
  NodeId tree_id = 1;
  NodeInfo node_info = 2;
  TreeStatus tree_status = 3;
}
